### ISSUE TITLE
[ENHANCEMENT]: Including cuco datastructures declarations for non-CUDA compilers.

### ISSUE THREAD

**Issue by dgabel** [NVIDIA/cuCollections#232]
### Is your feature request related to a problem? Please describe.

I would like to be able to have a declaration of the cuco datastructures available without having to compile the definition using CUDA so that they can be used in .hpp or .cpp files. At the moment, if you need a forward declaration you need to include all the default template arguments. The compiler doesn't allow redefinition of the default template arguments, even when they are the same as in the .cuh file.

### Describe the solution you'd like

Split the .cuh files into .hpp and an .inl. This way one can include either .hpp only for the declarations or .cuh for both the declarations and the definitions.

The .cuh file becomes:

#pragma once

#include <<data_structure>.hpp>
#include <cuco/detail/<data_structure>.inl>

I'm not sure this is that trivial as there may be cuda specific keywords in the declarations like __device__, __device__ and so on. 

Alternatively, the hpp file only contains the highest level declarations + default template value declarations. Then the only change to the current .cuh files would have to be to leave out these defaults in addition to the extra line:

#include <<data_structure>.hpp>

### Describe alternatives you've considered

Currently, forward declare the data structure with all its template arguments including those with default values and copy-paste their default values from the .cuh file into your declarations. Obviously, this is not ideal.

### Additional context

This is in particular convenient for larger codebases.

Any cpp file that depends on the cuco data structure through nested includes will run into this problem.

**Comment by jrhemstad** [comment]
I think we should be able to make it such that you could just do
```
#include <data_structure.cuh>
```
in a non-CUDA file and it would "just work" in that it wouldn't fail to compile so long as you don't try to do anything with the type.

With judicious use of [`__CUDACC__`](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-identification-macro), we should be able to make this work.

> `__CUDACC__`
>    Defined when compiling CUDA source files.

Similar to what @dgabel described, we can structure our `data_structure.cuh` headers like this:
```
//cuda.hpp

// make host/device annotations boil away in non-CUDA compilation
#if defined(__CUDACC__)
#define CUCO_HOST __host__
#define CUCO_DEVICE __device__ 
#define CUCO_HD __host__ __device__
#else
#define CUCO_HOST 
#define CUCO_DEVICE  
#define CUCO_HD 
#endif

// data_structure.cuh
#pragma once

struct data_structure{
   // declarations only
   auto foo(); 
   auto bar();
   CUCO_HD auto baz();
};

// Only include definitions of functions when compiling for CUDA
#if defined(__CUDACC__)
#include data_structure.inl
#endif
```

In other words, you can just include the same header as normal, and the header is careful to ensure CUDA-specific things aren't present when compiling without CUDA.

**Comment by jrhemstad** [comment]
As part of this we may want to move away from naming our headers with `.cuh` to make it more obvious that it is safe to include them in non-CUDA files.

**Comment by jrhemstad** [comment]
Alternatively, @davidwendt did something for `cudf::string_view` that is more similar to what @dgabel originally described with splitting into a [`string_view.hpp`](https://github.com/rapidsai/cudf/blob/branch-22.10/cpp/include/cudf/strings/string_view.hpp) and [`string_view.cuh`](https://github.com/rapidsai/cudf/blob/branch-22.10/cpp/include/cudf/strings/string_view.cuh) headers. 

For `cudf::string_view` there are legitimate use cases for using the data structure in non-CUDA code. I'm not sure if the same applies to cuco data structures.

**Comment by dgabel** [comment]
For me, just having the highest level data structure declarations would be sufficient.

It's not so much to have these available in cpp code, just to make sure the cpp code compiles if it relies on / interfaces with cuco CUDA code and data structures anywhere down the line.

**Comment by PointKernel** [comment]
We need to also make sure the dependent headers of `data_structure.cuh/hpp` (the example shown below) can work with non-CUDA compilers, i.e. separating declarations with implementations and guarding all implementations with `__CUDACC__` macro.
```cpp
#include <cuco/allocator.hpp>
#include <cuco/detail/__config>
#include <cuco/detail/error.hpp>
#include <cuco/detail/prime.hpp>
#include <cuco/probe_sequences.cuh>
#include <cuco/sentinel.cuh>
#include <cuco/traits.hpp>

#include <thrust/functional.h>

#include <cuda/std/atomic>

#if defined(CUCO_HAS_CUDA_BARRIER)
#include <cuda/barrier>
#endif

#include <cooperative_groups.h>
```

**Comment by PointKernel** [comment]
Based on the offline discussion, we need to check out how `thrust` handles this kind of issue.

**Comment by jrhemstad** [comment]
Another option is we could merge my idea and @dgabel and use `#if defined(__CUDACC__)` to conditionally just forward declare the type:

```
// Always forward declare
// Note: This will require all default parameters to also be forward declared types
template <typename Key,
          typename Value,
          cuda::thread_scope Scope = cuda::thread_scope_device,
          typename Allocator       = cuco::cuda_allocator<char>>
class static_map;

// Only do definition if compiling with CUDA
#if defined(__CUDACC__)

#include <cuco/allocator.hpp>
#include <cuco/detail/__config>
#include <cuco/detail/error.hpp>
#include <cuco/detail/hash_functions.cuh>
#include <cuco/detail/pair.cuh>
#include <cuco/detail/static_map_kernels.cuh>
#include <cuco/sentinel.cuh>
#include <cuco/traits.hpp>

#include <thrust/functional.h>

#include <cuda/std/atomic>

#if defined(CUCO_HAS_CUDA_BARRIER)
#include <cuda/barrier>
#endif

#include <cstddef>
#include <memory>
#include <utility>

namespace cuco {

template <typename Key, typename Value, cuda::thread_scope Scope, typename Allocator>
class dynamic_map;

template <typename Key,
          typename Value,
          cuda::thread_scope Scope,  // No defaults here anymore
          typename Allocator>
class static_map {
...
};

#include <cuco/detail/static_map.inl>

#endif

**Comment by sleeepyjack** [comment]
I am reviving this discussion because it was brought to my attention again today.

I don't think we can solve this issue easily for the legacy containers, but for the new `experimental::static_set/map` it should be doable.

The new design strictly splits the declaration from the definition, so @jrhemstad's first proposed solution is feasible.

My only concern is that renaming `.cuh` files to `.hpp` will directly affect any downstream projects that include these headers. It's the right thing to do, but we need to communicate the change to users before their code breaks.

**Comment by sleeepyjack** [comment]
This might be a stupid request, but can somebody come up with a simple unit test for this issue?

**Comment by mfbalin** [comment]
I am compiling my code for compute capability ranging from 35 to 90. Since `cuco::static_map` utilizes cuda::atomic inside of it, I can't even define a static_map in the host code when compiling for old+new architectures because `cuda/atomic` header checks the minimum compute capability and gives a compile time error when it is not greater than 60 or 70.
